README.md
Status IO [![NPM version][NPMIMGURL]][NPMURL]
===============
[NPMIMGURL]:                https://badge.fury.io/js/status-io.png
[NPM_INFO_IMG]:             https://nodei.co/npm/status-io.png?downloads=true&&stars
[NPMURL]:                   https://npmjs.org/package/status-io "npm"

[Status IO](http://coderaiser.github.io/status-io "Status IO"). Check host status.

Get the Status ![Status OK](http://status-io.cloudcmd.io/host/status-io.cloudcmd.io/ "Status")
===============
To get the status just drop this code to Markdown:

```
  ![Status IO](http://status-io.cloudcmd.io/host/host.io)
```

or Web Page.

```html
  <img alt="Status IO" src=http://status-io.cloudcmd.io/host/host.io />
```

You should change ```host.io``` to your host.

Install
===============
[![NPM_INFO][NPM_INFO_IMG]][NPMURL]

Statuses list
===============
When you set status image, you could see one of this posible statuses

- ![OK](http://status-io.cloudcmd.io/img/ok.png "OK") OK
- ![Moved](http://status-io.cloudcmd.io/img/moved.png "Moved") Moved or load to long
- ![Error](http://status-io.cloudcmd.io/img/error.png "Error") Error



ChangeLog
2018.10.24, v3.0.0

feature:
- (package) drop support of node < 8
- (package) files-io v3.0.0


2018.05.02, v2.0.1

fix:
- (service) start with nvm-exec
- (service) ExecStart

feature:
- (package) files-io v2.0.0


2017.10.06, v2.0.0

fix:
- (status-img) mime.lookup -> mime.getType

feature:
- (status-io) es2015-ify


2017.09.13, v1.0.2

fix:
- (sevice) v7.10.0

feature:
- (package) mime v2.0.0
- (service) nodejs v8.1.2


2016.11.03, v1.0.1

feature:
- (package) rm pipe-io
- (service) node v7
- (service) add


2015.12.07, v1.0.0

feature:
- (package) pipe-io: ~ -> ^
- (package) node: 0.6 -> 0.12
- (package) taus v1.1.0
- (package) mime v1.3.4
- (package) files-io v1.2.6
- (package) express v4.13.3


2015.08.17, v0.8.2

feature:
- feature(package) pipe-io v1.2.0


2015.06.04, v0.8.1

feature:
- (package) pipe-io v1.1.36


2014.11.20, v0.8.0

feature:
- (status) if response did not came in 1 second, send "moved"


2014.11.14, v0.7.3

feature:
- (package) add taus from npm


2014.11.14, v0.7.2

fix:
- (status) get back static


2014.11.14, v0.7.1

feature:
- (package) rm util-io


2014.11.14, v0.7.0

feature:
- (status-io) add lib



app.js
(function(){
    'use strict';
    
    var http        = require('http'),
        express     = require('express'),
        mime        = require('mime'),
        fs          = require('fs'),
        
        PORT        = 4321,
        OK          = 200,
        MOVED       = [301, 302],
        dir         = './img/',
        app         = express(),
        SVG         = 'svg',
        DIR         = SVG + '/',
        EXT         = '.' + SVG,
        OK_IMG      = DIR + 'ok'    + EXT,
        ERROR_IMG   = DIR + 'error' + EXT,
        MOVED_IMG   = DIR + 'moved' + EXT,
        TYPE        = mime.lookup(OK_IMG),
        ONE_SECOND  = 1000;
    
    http.createServer(app).listen(PORT);
    
    console.log('server: ' + PORT + '\npid: ' + process.pid);
   
    app.use('/', express.static(__dirname));
    
    app.get('/', function(req, res) {
        send(res, 'README.md', console.log.bind(console));
    });
    
    app.get('/host/*', function(request, response) {
        var host = 'http://' + request.params[0];
        
        console.log(request.params);
        
        if (host) {
            response.contentType(TYPE);
            
            request.setTimeout(ONE_SECOND, function() {
                send(response, MOVED_IMG);
            });
            
            http.get(host, function(res) {
                var status = res.statusCode;
                
                console.log(status);
                if (status === OK)
                    send(response, OK_IMG);
                else if(status === MOVED[0] || status === MOVED[1])
                    send(response, MOVED_IMG);
                else
                    send(response, ERROR_IMG);
            }).on('error', function(e) {
                response.contentType(TYPE);
                send(response, ERROR_IMG);
            });
        } else
            response.send('/:host');
    });
    
    function send(res, name, callback) {
        var read   = fs.createReadStream(name),
            error   = function (error) {
                res.send(error);
            },
            success = function () {
                if (typeof callback === 'function')
                    callback(name);
            };
        
        res.on('error', error);
        read.on('error', error);
        read.on('open', function() {
            read.pipe(res);
            read.on('end', success);
        });
    }
})();


lib/status-img.js
'use strict';

const path = require('path');
const mime = require('mime');
const files = require('files-io');

const status = require('taus');

const OK = 200;
const MOVED = [301, 302];
const DIR = path.normalize(__dirname + '/../');
const DIR_IMG = DIR + 'img/';
const EXT = '.png';
const IMG_OK = DIR_IMG + 'ok'    + EXT;
const IMG_ERROR = DIR_IMG + 'error' + EXT;
const IMG_MOVED = DIR_IMG + 'moved' + EXT;
const TYPE = mime.getType(IMG_OK);

const TIME = 1000;

module.exports = (addr, res) => {
    let sended;
    
    setTimeout(function() {
        if (!sended) {
            sended = true;
            sendFile(res, IMG_MOVED);
        }
    }, TIME);
    
    status(addr, (error, status) => {
        if (sended)
            return;
        
        sended = true;
        res.contentType(TYPE);
        
        if (status === OK)
            return sendFile(res, IMG_OK);
        
        if (status === MOVED[0] || status === MOVED[1])
            return sendFile(res, IMG_MOVED);
        
        sendFile(res, IMG_ERROR);
    });
};

function sendFile(res, name) {
    files.pipe(name, res).catch((error) => {
        res.send(error.message);
    });
}



status-ok.js
(function(){
    'use strict';
    
    var http        = require('http'),
        express     = require('express'),
        mime        = require('mime'),
        utilIO      = require('util-io'),
        pipeIO      = require('pipe-io'),
        
        PORT        = 4321,
        OK          = 200,
        MOVED       = [301, 302],
        app         = express(),
        DIR         = __dirname + '/img/',
        EXT         = '.png',
        OK_IMG      = DIR + 'ok'    + EXT,
        ERROR_IMG   = DIR + 'error' + EXT,
        MOVED_IMG   = DIR + 'moved' + EXT,
        TYPE        = mime.lookup(OK_IMG),
        TWO_SECONDS = 2000;
    
    http.createServer(app).listen(PORT);
    
    console.log('server: ' + PORT + '\npid: ' + process.pid);
    
    app.use('/', express.static(__dirname));
    
    app.get('/', function(req, res) {
        sendFile(res, 'README.md', console.log.bind(console));
    });
    
    app.get('/host/*', function(request, response) {
        var host = 'http://' + request.params[0],
            sended;
        
        console.log(request.params);
        
        if (host) {
            response.contentType(TYPE);
            
            setTimeout(function() {
                sended = true;
                sendFile(response, MOVED_IMG);
            }, TWO_SECONDS);
            
            http.get(host, function(res) {
                var status = res.statusCode;
                
                console.log(status);
                if (!sended)
                    if (status === OK)
                        sendFile(response, OK_IMG);
                    else if(status === MOVED[0] || status === MOVED[1])
                        sendFile(response, MOVED_IMG);
                    else
                        sendFile(response, ERROR_IMG);
                
                sended = true;
            }).on('error', function() {
                response.contentType(TYPE);
                sendFile(response, ERROR_IMG);
            });
        } else
            response.send('/:host');
    });
    
    function sendFile(res, name, callback) {
        pipeIO.create(name, res, function(error) {
            if (error)
                res.send(error, 404);
            else
                utilIO.exec(callback, name);
        });
    }
})();
